content/post/codesent.md

```diff
@@ -0,0 +1,395 @@
+
+We are going to use the following technologies:
+
+* [src-d/go-git](https://github.com/src-d/go-git) to read Git repositories.
+* [src-d/hercules](https://github.com/src-d/hercules) to analyze the commit tree.
+* [Babelfish](https://doc.bblf.sh) to extract the comments.
+* [vmarkovtsev/BiDiSentiment](https://github.com/vmarkovtsev/BiDiSentiment) to perform sentiment analysis (runs on Tensorflow).
+* Python to train the ML model and Go to apply it.
+
+I will describe each one in detail now.
+
+### go-git
+
+{{% caption src="/post/codesent/go-git.png" %}}
+go-git logo by Ricardo Baeta.
+{{% /caption %}}
+
+We started it back in 2015 as an "internal open source" project to fulfill our tasks of source code
+retrieval and analysis. This is an example of the typical software project workflow at source{d}
+today: we go open source from day one and see what happens next. This time go-git became a huge success
+with more than 2400 stars and several companies already using it in production. Although it lacks
+some of the [features](https://github.com/src-d/go-git/blob/master/COMPATIBILITY.md) of "big brothers",
+it is performant and highly extensible. Finally, the maintainers are experienced code maniacs and
+go-git is a good example of clean API and extensible architecture.
+
+### Hercules
+
+Hercules [appeared](https://blog.sourced.tech/post/hercules/) in December 2016 as an attempt
+to speed up Erik Bernhardsson's [Git-of-Theseus](https://erikbern.com/2016/12/05/the-half-life-of-code.html)
+by reimplementing it in Go and go-git. Indeed, it is an order of magnitude faster now.
+
+{{% caption src="/post/codesent/linux.png" %}}
+Line burndown of the Linux kernel - each band is one year and the thickness shows the ratio of lines
+which are not overwritten yet.
+{{% /caption %}}
+
+Then I added code ownership...
+
+{{% caption src="/post/codesent/emberjs_people.png" %}}
+Line ownership in Ember.js. Notice how one of devs appeared and conquered a half of the project.
+{{% /caption %}}
+
+...line overwrite matrix which shows how much code written by each developer is changed by others...
+
+{{% caption src="/post/codesent/wireshark_churn_matrix_black.png" %}}
+"Code wars" in Wireshark: XY cell's intensity is proportional to the number of times developer X's
+lines were overwritten with developer Y's.
+{{% /caption %}}
+
+...and even structural coupling - how tightly logically related the functions or classes are.
+
+{{% caption src="/post/codesent/jinja_black.png" %}}
+Structural clusters in Jinja2. Each dot is a function and each pair is closer proportionally to
+the number of times those two functions appeared in the same commit.
+{{% /caption %}}
+
+Under the cover, Hercules contains the DAG resolution engine which frees users from manually
+designing the analysis pipeline. It also leverages `pkg/plugin` to allow external extensions.
+
+### Babelfish
+
+We pass over to comment extraction. The most straightforward way to extract comments is to
```

Comment:
I would avoid using "pass over" (https://www.google.com/search?q=pass+over+meaning&oq=%22pass+over%22+&aqs=chrome.1.69i57j0l5.2102j0j1&sourceid=chrome&ie=UTF-8). And say "We continue with comment extraction."

Reply:
LOL. I want to kill my teachers now